VerticalWhitespaceClosingBracesRule: Add specialized configuration.

benjamin-kramer opened the pull request:
This solves issue #3940 

Replace the `SeverityConfiguration` with a new `VerticalWhitespaceClosingBracesConfiguration`, which includes `severity` and `only_enforce_before_trivial_lines`.

Vertical whitespace may be important for readability when the line with the closing brace is not a trivial one.
The configuration `only_enforce_before_trivial_lines` allows not enforcing the rule in such cases.

E.g.
```
if someCondition {
  // do something
  // do something
 
} else if ... {
  // do something
  // do something
  // do something

} else {
  // do something
  // do something
}
```

benjamin-kramer commented:
@jpsim Could you please assign someone to review this?

SimplyDanny commented on Source/SwiftLintFramework/Rules/Style/VerticalWhitespaceClosingBracesRule.swift:
```diff
@@ -8,121 +8,32 @@ private extension SwiftLintFile {
-        triggeringExamples: Array(violatingToValidExamples.keys).sorted(),
-        corrections: violatingToValidExamples.removingViolationMarkers()
+        nonTriggeringExamples: (examples.violatingToValidExamples.values +
+                                examples.nonTriggeringExamples).sorted(),
```
Why `.sorted()`?

benjamin-kramer replied:
I assumed this was the convention given the code that was previously present.
Removed (also in the next line).

SimplyDanny commented on Source/SwiftLintFramework/Rules/Style/VerticalWhitespaceClosingBracesRule.swift:
```diff
@@ -8,121 +8,32 @@ private extension SwiftLintFile {
+        nonTriggeringExamples: (examples.violatingToValidExamples.values +
+                                examples.nonTriggeringExamples).sorted(),
+        triggeringExamples: Array(examples.violatingToValidExamples.keys).sorted(),
+        corrections: examples.violatingToValidExamples.removingViolationMarkers()
```
No need to remove the violation markers.

benjamin-kramer replied:
Removed.

SimplyDanny commented on Tests/SwiftLintFrameworkTests/VerticalWhitespaceClosingBracesRuleTests.swift:
```diff
@@ -0,0 +1,75 @@
+import XCTest
+
+class VerticalWhitespaceClosingBracesRuleTests: XCTestCase {
+    func testVerticalWhitespaceClosingBracesWithDefaultConfiguration() {
```
Please remove this test. The examples are already verified in `GeneratedTests.swift`.

benjamin-kramer replied:
Removed.

SimplyDanny commented on Tests/SwiftLintFrameworkTests/VerticalWhitespaceClosingBracesRuleTests.swift:
```diff
@@ -0,0 +1,75 @@
+            """)
+    ]
+
+    func testMultilineArgumentsWithOnlyEnforceBeforeTrivialLines() {
```
Test name is not correct.

benjamin-kramer replied:
Oops. Renamed.

SimplyDanny commented on Tests/SwiftLintFrameworkTests/VerticalWhitespaceClosingBracesRuleTests.swift:
```diff
@@ -0,0 +1,75 @@
+            .with(nonTriggeringExamples: (nonTrivialLinesNonTriggeringExamples +
+                                          nonTrivialLinesViolatingToValidExamples.values))
+            .with(triggeringExamples: Array(nonTrivialLinesViolatingToValidExamples.keys))
+            .with(corrections: nonTrivialLinesViolatingToValidExamples.removingViolationMarkers())
```
No need to remove the violation markers.

benjamin-kramer replied:
Removed.

benjamin-kramer replied:
@SimplyDanny Just tried to run the tests.
They - both my own and the auto-generated ones - fail without the `.removingViolationMarkers()`.
What did you mean by this?

SimplyDanny replied:
In triggering examples, the markers show the position where the violation triggers. This is different in correction tests. In there, the locations mark the positions of applied fixes. Both can be different but are often the same. So to fix the test, the positions of the markers would either have to be changed or you remove them (as you did) in correction tests.

That said, you may ignore my comment. Sorry for the noise.

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -10,6 +10,10 @@
 
 #### Enhancements
 
+* VerticalWhitespaceClosingBracesRule: Add specialized configuration.  
```
Please be more descriptive here. I suggest something like:
```suggestion
* Add new option `only_enforce_before_trivial_lines` to 
  `vertical_whitespace_closing_braces` rule. It allows to
  restrict the rule to simple closing braces, while an empty
  line before closing braces followed by non-trivial code
  (if-statements, parameters, ...) will be accepted.  
```

benjamin-kramer replied:
Done.

SimplyDanny reviewed (changes requested):
Thank your for your contribution and your patience! I've added some remarks and questions.

Since I personally don't see a need for this option (I don't like empty lines before closing braces) and a discussion has not taken place in the issue, I would like to ask @jpsim for his feedback in addition.

Perhaps, you want to wait for the feedback before you consider my comments.

benjamin-kramer commented:
@SimplyDanny 
Made the corrections you suggested. Let me know whether to squash & rebase.

jpsim commented on Source/SwiftLintFramework/Rules/RuleConfigurations/VerticalWhitespaceClosingBracesConfiguration.swift:
```diff
@@ -0,0 +1,38 @@
+    case onlyEnforceBeforeTrivialLines = "only_enforce_before_trivial_lines"
+}
+
+// swiftlint:disable:next type_name
```
Instead of going for the long type name here, why not nest it in the `VerticalWhitespaceClosingBracesRule` type? E.g. `VerticalWhitespaceClosingBracesRule.Configuration`

benjamin-kramer replied:
Done.

benjamin-kramer replied:
Note that I moved the definition of the configuration from its own file into the `VerticalWhitespaceClosingBracesRule.swift` file, to avoid having to ignore additional rules (the filename and the extension of `VerticalWhitespaceClosingBracesRule` have different names).

jpsim commented:
> Let me know whether to squash & rebase.

Yes, please rebase.

jpsim commented on Tests/SwiftLintFrameworkTests/AutomaticRuleTests.generated.swift:
```diff
@@ -887,12 +887,6 @@ class VerticalWhitespaceBetweenCasesRuleTests: XCTestCase {
     }
 }
 
-class VerticalWhitespaceClosingBracesRuleTests: XCTestCase {
```
Let's keep the auto-generated tests. You can adjust which examples apply to which configurations by specifying the `configuration` parameter in the [`Example` initializer](https://realm.github.io/SwiftLint/Structs/Example.html#/s:18SwiftLintFramework7ExampleV_13configuration20testMultiByteOffsets0F17WrappingInComment0fjK6String0F14DisableCommand0F7OnLinux4file4line24excludeFromDocumentationACSS_ypSgS5bs06StaticM0VSuSbtcfc).

benjamin-kramer replied:
Done, though I'm not sure if it was done correctly.

SimplyDanny replied:
You can add the examples in the dedicated test class together with their configuration to the list of default examples. That way, they are automatically tested using the attached configuration.

benjamin-kramer replied:
Done.
Assuming the "dedicated test class" you mentioned was in `VerticalWhitespaceClosingBracesRuleTests.swift`, and "the list of default examples" is in `VerticalWhitespaceClosingBracesRuleExamples.swift`.

benjamin-kramer commented:
Fixed, squashed and rebased.

Note that I added the `.removingViolationMarkers()` again for the tests to pass.

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -357,6 +357,15 @@ macOS < 12.
+  the rule to apply only before trivial lines (containing
+  only closing braces, brackets and parentheses). This
+  allows empty lines before non-trivial lines of code
+  (e.g. if-else-statements).
```
This line is now missing the two trailing spaces:
```suggestion
  (e.g. if-else-statements).  
```
Sorry for the nitpicking. 😬

benjamin-kramer replied:
Done.

SimplyDanny commented on Source/SwiftLintFramework/Rules/Style/VerticalWhitespaceClosingBracesRuleExamples.swift:
```diff
@@ -0,0 +1,94 @@
+// swiftlint:disable:next type_name
+internal struct VerticalWhitespaceClosingBracesRuleExamples {
+    static let nonTriggeringExamples: [Example] = [
```
No need to have the explicit type in here:
```suggestion
    static let nonTriggeringExamples = [
```

benjamin-kramer replied:
Done.

SimplyDanny commented on Source/SwiftLintFramework/Rules/Style/VerticalWhitespaceClosingBracesRuleExamples.swift:
```diff
@@ -0,0 +1,94 @@
+        """)
+    ]
+
+    static let violatingToValidExamples: [Example: Example] = [
```
```suggestion
    static let violatingToValidExamples = [
```

benjamin-kramer replied:
Done.

benjamin-kramer commented:
Fixed, squashed and rebased.
I'm experiencing technical difficulties after the rebase (I can't manage to run the tests locally due to various missing packages), and some of the tests fail with seemingly unrelated problems.
Any idea what's going on here?

SimplyDanny commented:
Are you on Xcode 14/Swift 5.7?

benjamin-kramer commented:
Xcode 13.4, Swift 5.6.1.
In [this guide](https://github.com/realm/SwiftLint/blob/main/CONTRIBUTING.md), the xcode version seems somewhat out of date, and doesn't explain how to run just the tests.
When I try to do it (with a clean git status, and after deleting derived data) I get: 

> cannot find 'swiftparse_parser_set_language_version' in scope
> cannot find 'swiftparse_parser_set_enable_bare_slash_regex_literal' in scope

When trying the command line version I get:

> error: artifact of binary target 'lib_InternalSwiftSyntaxParser' has changed checksum; this is a potential security risk so the new artifact won't be downloaded
> error: artifact of binary target '_InternalSwiftSyntaxParser' has changed checksum; this is a potential security risk so the new artifact won't be downloaded

In any case, the CI system also has errors that seem unrelated to the changes in this PR.

SimplyDanny commented on Source/SwiftLintFramework/Rules/Style/VerticalWhitespaceClosingBracesRuleExamples.swift:
```diff
@@ -0,0 +1,147 @@
+        ↓
+        }])
+        """, configuration: beforeTrivialLinesConfiguration):
+            Example("""
```
> In any case, the CI system also has errors that seem unrelated to the changes in this PR.

This example needs `beforeTrivialLinesConfiguration` to be accepted. This is one (and perhaps the only) reason why CI fails.

benjamin-kramer replied:
I would never have guessed it from the errors...
Thanks!

benjamin-kramer commented:
Fixed and squashed.

SimplyDanny commented:
Thank you for all the (re-)work and your patience, @benjamin-kramer!

benjamin-kramer commented:
@SimplyDanny 
I just downloaded the latest release, that includes this PR.
When running `swiftlint rules` I don't see the configuration I added (nor is it mentioned [here](https://realm.github.io/SwiftLint/vertical_whitespace_closing_braces.html)). Why is this?

jpsim commented:
@benjamin-kramer 

When configuring rules, it needs to be done in a top-level yaml section:

```yaml
opt_in_rules:
  - vertical_whitespace_closing_braces

vertical_whitespace_closing_braces:
  only_enforce_before_trivial_lines: true
```

benjamin-kramer commented:
Thanks!
I verified that this works.
Doesn't explain the lack of documentation though.

jpsim commented:
> Doesn't explain the lack of documentation though.

Because the configuration description which is used to generate the docs is overridden to `N/A`: https://github.com/realm/SwiftLint/blob/60ad710b7f1ecb7edac8102cae0f2be6acf94f25/Source/SwiftLintFramework/Rules/Style/VerticalWhitespaceClosingBracesRule.swift#L60

I'll fix this shortly.

jpsim commented:
Fixing the docs in https://github.com/realm/SwiftLint/pull/4594
